core/amber/src/main/scala/edu/uci/ics/texera/web/model/event/WorkflowCompletedEvent.scala

```diff
@@ -26,32 +68,16 @@ object WorkflowCompletedEvent {
     val resultList = new mutable.MutableList[OperatorResult]
     workflowCompleted.result.foreach(pair => {
       val operatorID = pair._1
+      val chartType = OperatorResult.getChartType(operatorID, workflowCompiler)
 
-      // add chartType to result
-      val precedentOpID =
-        workflowCompiler.workflowInfo.links
-          .find(link => link.destination.operatorID == operatorID)
-          .get
-          .origin
-      val precedentOp =
-        workflowCompiler.workflowInfo.operators
-          .find(op => op.operatorID == precedentOpID.operatorID)
-          .get
-      val chartType = precedentOp match {
-        case operator: VisualizationOperator => operator.chartType()
-        case _                               => null
-      }
-
-      val table = precedentOp match {
-        case _: VisualizationOperator =>
-          pair._2.map(tuple => tuple.asInstanceOf[Tuple].asKeyValuePairJson())
-        case _ =>
+      val table = chartType match {
+        case Some(_) =>
           pair._2
-            .slice(0, defaultPageSize)
-            .map(tuple => tuple.asInstanceOf[Tuple].asKeyValuePairJson())
+        case None =>
+          pair._2.slice(0, defaultPageSize)
       }
 
-      resultList += OperatorResult(operatorID, table, chartType, pair._2.length)
+      resultList += OperatorResult.fromTuple(operatorID, table, chartType, pair._2.length)
```

Comment:
sorry it is not clear to me what is the type of pair._1 and pair._2, can we add some type declarations and comments here?

Reply:
fixed, now it's more readable